src/netconf_confirmed_commit.c

```diff
@@ -0,0 +1,824 @@
+/**
+ * @file netconf_confirmed_commit.c
+ * @author Tadeas Vintrlik <xvintr04@stud.fit.vutbr.cz>
+ * @brief ietf-netconf confirmed-commit capability callbacks
+ *
+ * @copyright
+ * Copyright 2021 Deutsche Telekom AG.
+ * Copyright (c) 2021 CESNET, z.s.p.o.
+ *
+ * This source code is licensed under BSD 3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#define _GNU_SOURCE
+
+#include "netconf_confirmed_commit.h"
+
+#include <assert.h>
+#include <dirent.h>
+#include <grp.h>
+#include <pwd.h>
+#include <signal.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+
+#include <libyang/libyang.h>
+#include <nc_server.h>
+#include <sysrepo.h>
+
+#include "common.h"
+#include "err_netconf.h"
+#include "log.h"
+
+#define RESTORE_FILE "/restore"
+
+/**
+ * @brief Permissions for Sysrepo
+ */
+typedef enum permissions {
+    R, /* Read */
+    W  /* Write */
+} perm_t;
+
+/**
+ * @brief Context for confirmed commits.
+ *
+ * When accessing struct members must use the lock mutex.
+ * Alternatively use the set_* and get_* functions
+ */
+typedef struct commit_ctx_s {
+    char *persist;  /* What persist-id is expected */
+    timer_t timer;        /* POSIX timer used for rollback, zero if none */
+    pthread_mutex_t lock; /* Lock mutexing this structure */
+} commit_ctx_t;
+
+static commit_ctx_t commit_ctx;
+
+/**
+ * @brief Initialize the commit_ctx_s structure.
+ */
+void
+ncc_commit_ctx_init(void)
+{
+    pthread_mutex_init(&commit_ctx.lock, NULL);
+    pthread_mutex_lock(&commit_ctx.lock);
+    commit_ctx.timer = NULL;
+    pthread_mutex_unlock(&commit_ctx.lock);
+}
```

Comment:
It is possible to write this as static initialization and thus remove the need for this function.